supoort calico v3

CMGS opened the pull request:
## Description

This PR make this plugin support calico v3, so we can use latest calicoctl manage now.  In this PR I do those things:

* use latest libcalico to implement ipam and network driver.
* set docker network endpoint in ippool's meta, thus we can find it without query docker api ( this will cause problem when system restart and plugin run in docker, both of docker and plugin were waiting each other)
* use hostname as default namespace, it's better when you have huge containers.
* tested fully and manually.
* I disable some tests because useless now. I think tests can cover whole cases.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/projectcalico/libnetwork-plugin?pullRequest=183) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/projectcalico/libnetwork-plugin?pullRequest=183) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/projectcalico/libnetwork-plugin?pullRequest=183) it.</sub>

CMGS commented:
@caseydavenport oh, sorry. because docker 17.05ce need `Sirupsen/logrus` I updated it to master now. And in my test, I found `fixed mac` not used in `cni-plugin`, thus I modified it. In PR you will see `CreateEndpoint` will not response `fixed mac address` any more. And when container join networks, after veth created, veth's mac will update to workloadendpoint in etcd like `cni-plugin`.

caseydavenport commented on driver/network_driver.go:
```diff
@@ -261,10 +291,12 @@ func (d NetworkDriver) CreateEndpoint(request *network.CreateEndpointRequest) (*
-	endpoint.Metadata.Orchestrator = d.orchestratorID
-	endpoint.Metadata.Workload = d.containerName
-	endpoint.Metadata.Name = request.EndpointID
+	endpoint.ObjectMeta.Namespace = hostname
```
Yup, that makes sense, but I think we should solve that more generally rather than overload the use of `namespace` - the wep already has a `node` field, so we can just make that an option in `calicoctl` for example - `calicoctl get weps --node mynode`

CMGS replied:
Yeah, that's sound good. But in this time calicoctl not support filter weps by node yet . And in data model I think it's a bit hard to implement. Because now we query workloads endpoint by query etcd with `prefix` (like /calico/v3/projectcalico.org/weps) and it will response all children with their data. So if we want use `calicoctl get weps --node somenode` that's means we have to **unmarshal** each data and filter in calicoctl with huge costs ( netio, cpu, memory ) if we have lots containers. However, if we set namespace with weps, it will store with `prefix/{NAMESPACE}` then etcd will help us to filter out what we need with less costs.

caseydavenport replied:
take a look at how the naming for libnetwork weps is meant to be: https://github.com/projectcalico/libcalico-go/blob/master/lib/names/workloadendpoint.go#L32

The first bit of the name is the node, and so lookup by node should be easily done at the etcd layer rather than by iteration.

CMGS replied:
But in etcd I found it store without node info like this
```
/calico/resources/v3/projectcalico.org/workloadendpoints/dev1/dev1-libnetwork-libnetwork-a7aba82dc07d2331001ed6d517e0403b6eba36865bc1f23d6c8601f79ec0287b

/calico/resources/v3/projectcalico.org/workloadendpoints/dev1/dev1-libnetwork-libnetwork-cda1849e824ab06255b4576df33749e5d86d1bc29cdde6f59f127cf753530806

/calico/resources/v3/projectcalico.org/workloadendpoints/slave/slave-libnetwork-libnetwork-98e3b4b92b03cc498ea86aa0ee7bcb1b192aebb15a2c881568a8abdd1a2eee9f
```

```dev1``` ```slave``` both set by namespace, not nodename

caseydavenport replied:
I think in your example above the format is like this:

```
/calico/resources/v3/projectcalico.org/workloadendpoints/<namespace>/<node>-libnetwork-libnetwork-<ID>
```

etcdv3 allows for prefix queries, so you can do a query against this:

```
/calico/resources/v3/projectcalico.org/workloadendpoints/<namespace>/<node>
```

To get all workloads within that namespace on that node.

caseydavenport replied:
I think we have a few options that come to mind:
- Use `libnetwork` as the namespace for everything.
- Use `libnetwork.<network_name>` as the namespace.
- Optionally, allow passing in a label to set the namespace.

For OpenStack, we just use `openstack` as the namespace, so just using `libnetwork` seems OK to me.

CMGS replied:
I think ```network_name``` and ```node``` both important for filtering weps in huge system. hmmm,  ```libnetwork.<network_name>``` seems better since node name was part of weps' name.

caseydavenport commented on tests/custom_if_prefix/libnetwork_env_var_test.go:
```diff
@@ -17,44 +18,48 @@ var _ = Describe("Running plugin with custom ENV", func() {
+			ip := dockerEndpoint.IPAddress
+			mac := dockerEndpoint.MacAddress
+			endpointID := dockerEndpoint.EndpointID
+			nicName := "cali" + endpointID[:mathutils.MinInt(11, len(endpointID))]
```
Strictly speaking, this isn't a `nic`, it's a `veth` :)

CMGS commented on tests/default_environment/libnetwork_test.go:
```diff
@@ -126,259 +122,164 @@ var _ = Describe("Libnetwork Tests", func() {
-			DockerString(fmt.Sprintf("docker rm -f %s", name))
-		})
-	})
+	//	Describe("docker run", func() {
```
OK, after my holiday I will fix it.

caseydavenport replied:
Have fun!

caseydavenport commented on tests/custom_if_prefix/libnetwork_env_var_test.go:
```diff
@@ -17,44 +18,48 @@ var _ = Describe("Running plugin with custom ENV", func() {
-				interface_name, mac, name, ip)))
+			key := fmt.Sprintf("/calico/resources/v3/projectcalico.org/workloadendpoints/%s/%s-libnetwork-libnetwork-%s", "libnetwork", pool, endpointID)
+			endpointJSON := GetEtcd(key)
+			wep := map[string]interface{}{}
```
I think you can use libcalico structs here - e.g. unmarshal wep into an `v3.WorkloadEndpoint` struct.

And, if you're up for it, we could ditch `GetEtcd` altogether and use libcalico to query etcd here. That would clean this code up a lot. However, I don't think we _need_ to do that to get this merged.

caseydavenport commented on tests/custom_if_prefix/libnetwork_env_var_test.go:
```diff
@@ -17,44 +18,48 @@ var _ = Describe("Running plugin with custom ENV", func() {
-			Expect(tags).Should(MatchJSON(fmt.Sprintf(`["%s"]`, name)))
-			Expect(labels).Should(MatchJSON("{}"))
-			Expect(rules).Should(MatchJSON(fmt.Sprintf(`{"inbound_rules": [{"action": "allow","src_tag": "%s"}],"outbound_rules":[{"action": "allow"}]}`, name)))
+			//// Check profile
```
nit: two extra `//`

caseydavenport commented on tests/custom_if_prefix/libnetwork_env_var_test.go:
```diff
@@ -17,44 +18,48 @@ var _ = Describe("Running plugin with custom ENV", func() {
+			profile := map[string]interface{}{}
+			json.Unmarshal(profileJSON, &profile)
+			meta := profile["metadata"].(map[string]interface{})
+			Expect(meta["name"].(string)).Should(Equal(pool))
```
It seems we've lost the assertion that the tags / labels / rules are correct here. Could you add those checks back?

caseydavenport commented on tests/custom_wep_labelling/libnetwork_env_var_test.go:
```diff
@@ -17,75 +18,43 @@ var _ = Describe("Running plugin with custom ENV", func() {
-		})
-	})
-
-	Describe("docker run", func() {
```
Shouldn't we keep this test?

caseydavenport commented on tests/custom_wep_labelling/libnetwork_env_var_test.go:
```diff
@@ -17,75 +18,43 @@ var _ = Describe("Running plugin with custom ENV", func() {
+			profile := map[string]interface{}{}
+			json.Unmarshal(profileJSON, &profile)
+			meta := profile["metadata"].(map[string]interface{})
+			Expect(meta["name"].(string)).Should(Equal(pool))
```
Similar here - we've lost the assertion on tags / labels / rules.

CMGS commented:
@caseydavenport tests added. Subnet test no need anymore cause calico v3 need create network with subnet now.

caseydavenport commented on driver/network_driver.go:
```diff
@@ -76,6 +78,11 @@ func NewNetworkDriver(client *datastoreClient.Client) network.Driver {
 		labelEndpoints: strings.EqualFold(os.Getenv(LABEL_ENDPOINTS_ENVKEY), "true"),
 	}
 
+	ns := os.Getenv(NAMESPACE_ENVKEY)
```
Won't this configure the namespace for all users of this driver? I don't think that's right - I don't think we want a per-node namespace configuration option.

caseydavenport replied:
Could we use a container label instead?

CMGS replied:
oh! I misunderstand you! However use container label will query docker in create/join/delete, I think that's will make plugin require docker's response. When docker daemon restarting, it will cause block problems. I think ```libnetwork.<network_name>``` will be better, I will change it on Monday.

caseydavenport commented on driver/network_driver.go:
```diff
@@ -205,9 +214,37 @@ func (d NetworkDriver) CreateNetwork(request *network.CreateNetworkRequest) erro
+				ipPool.SetAnnotations(ann)
+				_, err = poolClient.Update(ctx, &ipPool, options.SetOptions{})
+				if err != nil {
+					log.Errorln(err)
```
Shouldn't we return an error here?

caseydavenport commented on driver/network_driver.go:
```diff
@@ -574,25 +657,31 @@ RETRY_CONTAINER_INSPECT:
 		return
 	}
 
+	rev, err := strconv.Atoi(endpoint.ObjectMeta.ResourceVersion)
```
I don't think this is right - could you remove the bit that updates the resource version?

CMGS replied:
@tmjd  however, it will failed when updating with same rev if CALICO_LIBNETWORK_LABEL_ENDPOINT set true.

```ERRO[0086] Unable to update WorkloadEndpoint with labels (T=400.626763ms): update conflict: WorkloadEndpoint(libnetwork/slave-libnetwork-libnetwork-c8f7152c607b86720cf117bf95a3cb3b00c02d2b05215e2859cdab8004735e45)```

thus means we have to update revision.

caseydavenport replied:
I think that's because when CALICO_LIBNETWORK_LABEL_ENDPOINT is true, there are two goroutines which are attempting to udpate the workload endpoint, and so one of them will fail (because it will have an old revision).

Rather than manually adjust the revision, we should use a retry loop for this case.

We do something along these lines in other similar bits of code:

```
for range(10) {
  wep, err := Get(wep)
  if err != nil {
    break
  }

  err := Update(wep)
  if err != nil && err == UpdateConflict {
    // Conflict updating WEP - retry
    time.Sleep(10ms)
    continue
  }
break
}
```

CMGS commented on driver/network_driver.go:
```diff
@@ -260,68 +298,89 @@ func (d NetworkDriver) CreateEndpoint(request *network.CreateEndpointRequest) (*
-	endpoint.Metadata.Name = request.EndpointID
+	endpoint.Name = wepName
+	//endpoint.ObjectMeta.Namespace = fmt.Sprintf("%s.%s", d.orchestratorID, networkName)
+	endpoint.ObjectMeta.Namespace = d.namespace
```
I think it will be a bit hard for getting network_name when join/delete a endpoint. Docker call those interfaces with ID, we have to query docker daemon. In my opinion cyclic call is not a good idea. So let's keep namespace as libnetwork.

caseydavenport replied:
sgtm

CMGS commented:
I don't know why Update will failed when testing. Is there any options for me to get plugin output when testing?

mountkin commented on driver/ipam_driver.go:
```diff
@@ -92,16 +94,31 @@ func (i IpamDriver) RequestPool(request *ipam.RequestPoolRequest) (*ipam.Request
-		pool = request.Pool
-		poolID = request.Pool
+
+		fmt.Println(pool, poolID)
```
I think this line should be removed.

CMGS commented:
@caseydavenport @mountkin done

caseydavenport commented:
Awesome, thanks for all of the hard work on this!
